Sources/MongoSwift/BSON/Document+Sequence.swift

```diff
@@ -10,13 +10,156 @@ import libmongoc
+            return DocumentIterator.subsequence(of: self, startIndex: n)
+        }
+    }
+
+    public func dropLast(_ n: Int) -> Document {
+        switch n {
+        case ..<0:
+            preconditionFailure("Can't drop a negative number of elements from a `Document`")
+        case 0:
+            return self
+        default:
+            // the subsequence we want is [0, length - n)
+            let end = self.count - n
+            // if we are dropping >= the length, just short circuit and return empty doc
+            return end <= 0 ? [:] : DocumentIterator.subsequence(of: self, endIndex: end)
+        }
+    }
+
+    public func drop(while predicate: (KeyValuePair) throws -> Bool) rethrows -> Document {
+        // tracks whether we are still in a "dropping" state. once we encounter
+        // an element that doesn't satisfy the predicate, we stop dropping.
+        var drop = true
+        return try self.filter { elt in
+            if drop {
+                // still in "drop" mode and it matches predicate
+                if try predicate(elt) { return false }
+                // else we've encountered our first non-matching element
+                drop = false
+                return true
+            }
+            // out of "drop" mode, so we keep everything
+            return true
+        }
+    }
+
+    public func prefix(_ maxLength: Int) -> Document {
+        switch maxLength {
+        case ..<0:
+            preconditionFailure("Can't retrieve a negative length prefix of a `Document`")
+        case 0:
+            return [:]
+        default:
+            // short circuit if there are fewer elements in the doc than requested
+            return self.count <= maxLength ? self : DocumentIterator.subsequence(of: self, endIndex: maxLength)
+        }
+    }
+
+    public func prefix(while predicate: (KeyValuePair) throws -> Bool) rethrows -> Document {
+        var output = Document()
+        for elt in self {
+            if try !predicate(elt) { break }
+            output[elt.key] = elt.value
+        }
+        return output
+    }
+
+    public func suffix(_ maxLength: Int) -> Document {
+        switch maxLength {
+        case ..<0:
+            preconditionFailure("Can't retrieve a negative length suffix of a `Document`")
+        case 0:
+            return [:]
+        default:
+            let start = self.count - maxLength
+            // short circuit if there are fewer elements in the doc than requested
+            return start <= 0 ? self : DocumentIterator.subsequence(of: self, startIndex: start)
+        }
+    }
+
+    public func split(maxSplits: Int = Int.max, omittingEmptySubsequences: Bool = true,
+                      whereSeparator isSeparator: (KeyValuePair) throws -> Bool) rethrows -> [Document] {
+        // rather than implementing the complex logic necessary for split, convert to an array and call split on that
```

Comment:
for several methods besides `split` added in this PR, I wrote custom implementations relying on a new static method `DocumentIterator.subsequence`. my reasoning was that several of those methods allow us to get away with only reading out some portion of the `Document` into Swift types --  we often get to skip over values at the front of the document or terminate early. but calling `Array(self)` and then doing the operations on that array would always requires us to read every single value out to the array, which isn't very efficient. 
On the other hand, `split` is always going to require reading every single element anyway, plus the source code for the array version is pretty complicated (see [here](https://github.com/apple/swift/blob/a17dbfbf29aeed3f80b484a5a66c9ea5b48af3d3/stdlib/public/core/Collection.swift#L1652)), so that was another compelling reason to just do this rather than trying to duplicate.